ev3dev/motor.py

```diff
@@ -1695,3 +1699,199 @@ def on_for_seconds(self, steering, speed_pct, seconds, brake=True, block=True):
     def on(self, steering, speed_pct):
         (left_speed_pct, right_speed_pct) = self.get_speed_steering(steering, speed_pct)
         MoveTank.on(self, left_speed_pct, right_speed_pct)
+
+
+class MoveJoystick(MoveTank):
+    """
+    Used to control a pair of motors via a joystick
+    """
+
+    def angle_to_speed_percentage(self, angle):
+        """
+                                (1, 1)
+                             . . . . . . .
+                          .        |        .
+                       .           |           .
+              (0, 1) .             |             . (1, 0)
+                   .               |               .
+                  .                |                 .
+                 .                 |                  .
+                .                  |                   .
+               .                   |                   .
+               .                   |     x-axis        .
+       (-1, 1) .---------------------------------------. (1, -1)
+               .                   |                   .
+               .                   |                   .
+                .                  |                  .
+                 .                 | y-axis          .
+                   .               |               .
+             (0, -1) .             |             . (-1, 0)
+                       .           |           .
+                          .        |        .
+                             . . . . . . .
+                                (-1, -1)
+
+
+        The joystick is a circle within a circle where the (x, y) coordinates
+        of the joystick form an angle with the x-axis.  Our job is to translate
+        this angle into the percentage of power that should be sent to each motor.
+        For instance if the joystick is moved all the way to the top of the circle
+        we want both motors to move forward with 100% power...that is represented
+        above by (1, 1).  If the joystick is moved all the way to the right side of
+        the circle we want to rotate clockwise so we move the left motor forward 100%
+        and the right motor backwards 100%...so (1, -1).  If the joystick is at
+        45 degrees then we move apply (1, 0) to move the left motor forward 100% and
+        the right motor stays still.
+
+        The 8 points shown above are pretty easy. For the points in between those 8
+        we do some math to figure out what the percentages should be. Take 11.25 degrees
+        for example. We look at how the motors transition from 0 degrees to 45 degrees:
+        - the left motor is 1 so that is easy
+        - the right motor moves from -1 to 0
+
+        We determine how far we are between 0 and 45 degrees (11.25 is 25% of 45) so we
+        know that the right motor should be 25% of the way from -1 to 0...so -0.75 is the
+        percentage for the right motor at 11.25 degrees.
+        """
+
+        if angle >= 0 and angle <= 45:
```

Comment:
Is there really no less verbose way of doing this?

Reply:
Not that I can think of off the top of my head.

Reply:
Looks like `0 <= angle <= 45` should work:
https://stackoverflow.com/questions/13628791/how-do-i-check-whether-an-int-is-between-the-two-numbers

Reply:
oh I thought wasabi was asking if the entire function could be made less verbose but if it is just the angle comparisons yep I can clean that up to a single operation for each if/elif

Reply:
May be you are right. I tried to plot the left and right motor powers from your comment against the angle to see if there is a simple mathematical expression lurking there, but it makes for some weirdly interesting plots.

Reply:
I remember when I was working on this I would take the joystick and move it around the outer edge of the circle.  Initially I got some really crazy movements but then figured out that there are 8 sections of the circle to deal with thus the big if/elif tree. How the motor speeds adjust as you move from one section to the next varies a lot depending on which two sections you are dealing with.

Reply:
The plots: http://nbviewer.ipython.org/745ed4980597830194766f8e865dba69

Reply:
Here is the plot for old `steering` function.
http://nbviewer.jupyter.org/gist/anonymous/d2750f5923b3523f8c5e25812adb7a20
It does look the same as the top part of your circle, which I think is a good sign. Not sure though that the bottom part of your circle is correct. Should not there be some kind of symmetry?

Reply:
Nah, I think your function is correct.

Reply:
Cool plots

Reply:
Does this make more sense?
http://nbviewer.jupyter.org/gist/anonymous/bc23dcefc6da4f0008df620d33e4d171

Reply:
I would have to build a tank robot and test it to say for sure but my gut says no just because it looks so different from what I have and I know that what I have works (and took me a while to get right).  It would be worth testing though.

Reply:
I'll test it out sometime in the next week....I just built MINDCUBER yesterday so all of my large motors are tied up right now.  Anyone know of a tank base that uses medium motors?